modules/mod-deep-learning/DeepLearningEffectBase.cpp

```diff
@@ -0,0 +1,363 @@
+std::vector<ClipTimestamps> DeepLearningEffectBase::GetClipTimestamps(WaveTrack *track, double tStart, double tEnd) const
+{
+   std::vector<ClipTimestamps> timestamps;
+
+   for (const auto &clip : track->GetClips())
+      timestamps.emplace_back(
+         std::pair<double, double>(clip->GetPlayStartTime(), 
+                                   clip->GetPlayEndTime())
+      );
+
+   return timestamps;
+}
+
+
+torch::Tensor DeepLearningEffectBase::BuildMonoTensor(WaveTrack *track, float *buffer,
+                                                  sampleCount start, size_t len)
+{
+   //Get the samples from the track and put them in the buffer
+   if (!track->GetFloats(buffer, start, len))
+      throw std::runtime_error("An error occurred while copying samples to tensor buffer.");
+
+   // get tensor input from buffer
+   torch::Tensor audio = torch::from_blob(buffer, len,
+                                          torch::TensorOptions().dtype(torch::kFloat));
+   audio = audio.unsqueeze(0); // add channel dimension
+
+   return audio;
+}
+
+torch::Tensor DeepLearningEffectBase::BuildMultichannelTensor(WaveTrack *leader, float *buffer,
+                                                          sampleCount start, size_t len)
+{
+   auto channels = TrackList::Channels(leader);
+   std::vector<torch::Tensor> channelStack;
+
+   // because we're reusing the same buffer, it's important that
+   // we clone the tensor.
+   for (WaveTrack *channel : channels)
+      channelStack.emplace_back(
+          BuildMonoTensor(channel, buffer, start, len).clone());
+
+   return torch::cat(channelStack);
+}
+
+torch::jit::IValue DeepLearningEffectBase::ForwardPassInThread(torch::Tensor input)
+{
+   torch::jit::IValue output;
+
+   std::atomic<bool> done = {false};
+   std::atomic<bool> success = {true};
+
+   // make a copy of the model (in case we need to abort)
+   DeepModelHolder model = this->mModel;
+
+   auto thread = std::thread(
+      [model, &input, &output, &done, &success]()
+      {
+         try
+         {
+            torch::jit::IValue tempOut = model->Forward(input);
+
+            // only write to output tensor if abort was not requested
+            if (success)
+               output = tempOut;
+         }
+         catch (const ModelException &e)
+         {
+            wxLogError(e.what());
+            wxLogDebug(e.what());
+            success = false;
+            output = torch::jit::IValue(torch::zeros_like(input));
+         }
+         done = true;
+      }
+   );
+
+   // wait for the thread to finish
+   while (!done)
+   {
+      if (TrackProgress(mCurrentTrackNum, mCurrentProgress))
+      {
+         // abort if requested
+         success = false;
+
+         // tensor output will be destroyed once the thread is destroyed
+         thread.detach();
+
+         output = torch::jit::IValue(torch::zeros_like(input));
+         return output;
+      }
+
+      ::wxSafeYield();
+      wxMilliSleep(50);
+   }
+   thread.join();
```

Comment:
Precede this line with
`if (thread.joinable())`
else, a std:: exception will be thrown in the case that the thread was detached.